docs/structured-streaming-kafka-integration.md

```diff
@@ -0,0 +1,231 @@
+
+</div>
+<div data-lang="java" markdown="1">
+
+    // Subscribe to 1 topic
+    Dataset<Row> ds1 = spark
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribe", "topic1")
+      .load()
+    ds1.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+    // Subscribe to multiple topics
+    Dataset<Row> ds2 = spark
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribe", "topic1,topic2")
+      .load()
+    ds2.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+    // Subscribe to a pattern
+    Dataset<Row> ds3 = spark
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribePattern", "topic.*")
+      .load()
+    ds3.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+</div>
+<div data-lang="python" markdown="1">
+
+    # Subscribe to 1 topic
+    ds1 = spark
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribe", "topic1")
+      .load()
+    ds1.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+    # Subscribe to multiple topics
+    ds2 = spark
+      .readStream
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribe", "topic1,topic2")
+      .load()
+    ds2.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+    # Subscribe to a pattern
+    ds3 = spark
+      .readStream()
+      .format("kafka")
+      .option("kafka.bootstrap.servers", "host1:port1,host2:port2")
+      .option("subscribePattern", "topic.*")
+      .load()
+    ds3.selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+
+</div>
+</div>
+
+Each row in the source has the following schema:
+<table class="table">
+<tr><th>Column</th><th>Type</th></tr>
+<tr>
+  <td>key</td>
+  <td>binary</td>
+</tr>
+<tr>
+  <td>value</td>
+  <td>binary</td>
+</tr>
+<tr>
+  <td>topic</td>
+  <td>string</td>
+</tr>
+<tr>
+  <td>partition</td>
+  <td>int</td>
+</tr>
+<tr>
+  <td>offset</td>
+  <td>long</td>
+</tr>
+<tr>
+  <td>timestamp</td>
+  <td>long</td>
+</tr>
+<tr>
+  <td>timestampType</td>
+  <td>int</td>
+</tr>
+</table>
+
+Right now, the Kafka source has the following Spark's specific options.
+
+<table class="table">
+<tr><th>Option</th><th>value</th><th>default</th><th>meaning</th></tr>
+<tr>
+  <td>startingOffset</td>
+  <td>["earliest", "latest"]</td>
+  <td>"latest"</td>
+  <td>The start point when a query is started, either "earliest" which is from the earliest offset, 
+  or "latest" which is just from the latest offset. Note: This only applies when a new Streaming q
+  uery is started, and that resuming will always pick up from where the query left off.</td>
+</tr>
+<tr>
+  <td>failOnCorruptMetadata</td>
```

Comment:
There are 3 places calling `reportCorruptMetadata`:
- We find some new partitions, then try to connect to query their earliest offsets but these partitions are gone
  - If these partitions are empty, there is no data lost.
- Find new partitions but their earliest offsets are not 0.
- Deleting a topic and create the some topic. This makes the new offset less that the old one.
  - There may be no data lost if we have already processed all data in the old topic.